File: backend/app/models/notebook.py
import enum
from datetime import datetime
from sqlalchemy import Column, String, Integer, SmallInteger, Boolean, DateTime, Text, ForeignKey, Enum
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.models.base import generate_cuid


class NoteType(str, enum.Enum):
    STEP = "step"           # Time management - free step
    TASK = "task"           # Time management - task with status
    IDEA = "idea"           # Ideas section


class TaskStatus(str, enum.Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    DONE = "done"


class NotebookTopic(Base):
    __tablename__ = "nb_topics"

    id = Column(String, primary_key=True, default=generate_cuid)
    user_id = Column(String, nullable=False)  # telegram chat_id or web user
    name = Column(String(100), nullable=False)
    description = Column(String(500), nullable=True)
    emoji = Column(String(10), nullable=True)
    is_predefined = Column(Boolean, default=False)
    is_deleted = Column(Boolean, default=False)
    created_at = Column(DateTime, default=datetime.utcnow)

    notes = relationship("NotebookNote", back_populates="topic")


class NotebookNote(Base):
    __tablename__ = "nb_notes"

    id = Column(String, primary_key=True, default=generate_cuid)
    user_id = Column(String, nullable=False)
    note_type = Column(String(20), nullable=False)  # step, task, idea
    topic_id = Column(String, ForeignKey("nb_topics.id"), nullable=True)
    content = Column(Text, nullable=False)
    step_order = Column(SmallInteger, nullable=True)
    task_status = Column(String(20), nullable=True)  # todo, in_progress, done
    is_deleted = Column(Boolean, default=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    topic = relationship("NotebookTopic", back_populates="notes")
    history = relationship("NotebookNoteHistory", back_populates="note", order_by="NotebookNoteHistory.edited_at.desc()")


class NotebookNoteHistory(Base):
    __tablename__ = "nb_note_history"

    id = Column(String, primary_key=True, default=generate_cuid)
    note_id = Column(String, ForeignKey("nb_notes.id", ondelete="CASCADE"), nullable=False)
    content = Column(Text, nullable=False)
    edited_at = Column(DateTime, default=datetime.utcnow)

    note = relationship("NotebookNote", back_populates="history")

File: frontend/src/features/notebook/api/notebook.ts
import client from '../../../shared/api/client';

export interface Topic {
  id: string;
  name: string;
  emoji: string | null;
  description: string | null;
  isPredefined: boolean;
  ideaCount: number;
  createdAt: string | null;
}

export interface Note {
  id: string;
  noteType: 'step' | 'task' | 'idea';
  topicId: string | null;
  content: string;
  stepOrder: number | null;
  taskStatus: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

export const notebookApi = {
  // Topics
  getTopics: () => client.get<Topic[]>('/notebook/topics').then((r) => r.data),
  createTopic: (data: { name: string; emoji?: string; description?: string }) =>
    client.post<Topic>('/notebook/topics', data).then((r) => r.data),
  updateTopic: (id: string, data: { name?: string; emoji?: string; description?: string }) =>
    client.put<Topic>(`/notebook/topics/${id}`, data).then((r) => r.data),
  deleteTopic: (id: string) => client.delete(`/notebook/topics/${id}`),

  // Steps
  getSteps: () => client.get<Note[]>('/notebook/steps').then((r) => r.data),
  addStep: (content: string) =>
    client.post<Note>('/notebook/steps', { content }).then((r) => r.data),

  // Tasks
  getTasks: (status?: string) =>
    client.get<Note[]>('/notebook/tasks', { params: status ? { status } : {} }).then((r) => r.data),
  addTask: (content: string, taskStatus?: string) =>
    client.post<Note>('/notebook/tasks', { content, taskStatus }).then((r) => r.data),

  // Ideas
  getIdeas: (topicId: string) =>
    client.get<Note[]>(`/notebook/ideas/${topicId}`).then((r) => r.data),
  addIdea: (topicId: string, content: string) =>
    client.post<Note>(`/notebook/ideas/${topicId}`, { content }).then((r) => r.data),

  // Notes (edit/delete any type)
  updateNote: (id: string, data: { content?: string; taskStatus?: string }) =>
    client.put<Note>(`/notebook/notes/${id}`, data).then((r) => r.data),
  deleteNote: (id: string) => client.delete(`/notebook/notes/${id}`),
};
